nncf/quantization/passes.py

```diff
@@ -34,27 +36,33 @@ def transform_to_inference_graph(
     :param dropout_metatypes: List of backend-specific Dropout metatypes.
     :return: NNCFGraph in the inference style.
     """
-    remove_shapeof_subgraphs(nncf_graph, shapeof_metatypes, input_nodes)
-    filter_constant_nodes(nncf_graph, input_nodes)
+    shapeof_subgraphs = find_shapeof_subgraphs(nncf_graph, shapeof_metatypes, input_nodes)
+    preserved_nodes = find_preserved_nodes(nncf_graph, shapeof_subgraphs, preserved_metatypes)
+    constant_subgraphs = find_constant_subgraphs(nncf_graph, input_nodes)
+
+    nodes_to_drop = set([*shapeof_subgraphs, *constant_subgraphs]).difference(preserved_nodes)
+    nncf_graph.remove_nodes_from(nodes_to_drop)
+
     remove_nodes_and_reconnect_graph(nncf_graph, dropout_metatypes)
     return nncf_graph
 
 
-def remove_shapeof_subgraphs(
+def find_shapeof_subgraphs(
```

Comment:
It looks like this function does not find shape of subgraph, but find all nodes after shapeof operation until nodes which have path to an input node.

Let discuss algorithm offline.

Reply:
The goal of follow-up PR generalize algorithm for all quantizable operations.

Reply:
ref: 150942